### DISCUSSION TITLE
La red aprende solo lo que vé

### DISCUSSION THREAD

**Discussion by josePereiro** [josePereiro/TankeFactory#12] [Discussions]
[Andrej Karpathy](https://github.com/karpathy) En este [video](https://www.youtube.com/watch?v=PaCmpygFfXo) hace una red que aprende de los datos la matrix $P(c1, c2)$. O sea, la frecuencia de cada par de caracteres en un texto origen (Shakespeare stuff).

De modo generativo, uno puede usar $P(c1, c2)$ para generar un $c2$ a partir de un $c1$. O sea, dado el caracter donde estoy dame el siguiente. Formalmente creo se puede definir como $P_g(c2|c1) \approx P(c1, c2)$. Donde $P_g$ es la distribución generadora. Fijense que, por construcción, el texto generado va ha tener la misma matriz de frecuencia que el original...

Ahora, uno puede diseñar una red para que sea dicho generador. Eso es lo que hace Karpathy, por eso le llama BiGramModel... Un texto generado con una red donde para predecir el siguiente caracter/token solo se mire del pasado un caracter/token, lo único que compartirá con el texto original es que ambos tendrán la misma matriz de frecuencia para cada par de caracteres/tokens. Este resultado, de ser alcanzable, debe ser independiente de la arquitectura.

Una cuestión interesante es la diferencia entre: 
 i) un modelo donde para generar un token se mire dos del pasado,  y 
ii) un modelo donde para generar dos tokens se mire uno. 

Los dos modelos usarían el mismo set de datos de entrenamiento, triples que aparecerán cada uno con una frecuencia dada.
Los dos estarían reproduciendo solo la matriz de frecuencia de los tríos de tokens? Un TriGramModel?

Noten que por definición si tienes P(c1,c2,c3) tienes P(c1,c2) y P(c1)...

**Comment by Fourier92** [comment]
> [Andrej Karpathy](https://github.com/karpathy) En este [video](https://www.youtube.com/watch?v=PaCmpygFfXo) hace una red que aprende de los datos la matrix P(c1,c2). O sea, la frecuencia de cada par de caracteres en un texto origen (Shakespeare stuff).
> 
> De modo generativo, uno puede usar P(c1,c2) para generar un c2 a partir de un c1. O sea, dado el caracter donde estoy dame el siguiente. Formalmente creo se puede definir como Pg(c2|c1)≈P(c1,c2). Donde Pg es la distribución generadora. Fijense que, por construcción, el texto generado va ha tener la misma matriz de frecuencia que el original...
> 
> Ahora, uno puede diseñar una red para que sea dicho generador. Eso es lo que hace Karpathy, por eso le llama BiGramModel... Un texto generado con una red donde para predecir el siguiente caracter/token solo se mire del pasado un caracter/token, lo único que compartirá con el texto original es que ambos tendrán la misma matriz de frecuencia para cada par de caracteres/tokens. Este resultado, de ser alcanzable, debe ser independiente de la arquitectura.
> 
> Una cuestión interesante es la diferencia entre: i) un modelo donde para generar un token se mire dos del pasado, y ii) un modelo donde para generar dos tokens se mire uno.
> 
> Los dos modelos usarían el mismo set de datos de entrenamiento, triples que aparecerán cada uno con una frecuencia dada. Los dos estarían reproduciendo solo la matriz de frecuencia de los tríos de tokens? Un TriGramModel?
> 
> Noten que por definición si tienes P(c1,c2,c3) tienes P(c1,c2) y P(c1)...

Esos modelos a que haces referencia estarían reproduciendo diferentes matrices de probabilidad. Pensemos en este ejemplo: tienes una matriz con la frecuencia de aparición de dos caracteres, dado que han ocurrido otros dos, $P(c3, c4|c1,c2)$ sería una matriz donde las filas son todos las posibles combinaciones de dos caracteres y las columnas lo mismo, el elemento $[i,j]$ sería la probabilidad condicional de que dado que ocurrió esta combinación específica de dos caracteres en la fila $i$, ocurre esta otra combinación de la columna $j$.
Basados en esta matriz se pueden construir ambos casos que mencionabas. El primero dado dos caracteres, predecir el próximo, bastaría ver los últimos dos caracteres encontrarlos en la fila $k$ y luego de todas las diferentes opciones de dos elementos que aparecen en la fila $k$ sumar todos los empiecen con el mismo carácter, para todos los caracteres, y eso dará la probabilidad del próximo carácter. Al samplear texto basado en esto, se obtiene una instancia que sigue la probabilidad condicional $P(c3|c1, c2)$, no la original $P(c3,c4|c1,c2)$.
En el segundo caso que refieres, dado un carácter samplear los próximos dos: basado en la misma matriz del inicio, buscamos en las filas todos los dúos que terminen con el carácter en cuestión y tenemos que sumar todos los vectores de probabilidades de estas filas, normalizar y el caso con más alta probabilidad resultante es el candidato a ser sampleado, y es esta nueva matriz de probabilidad la que estaríamos sampleado: $P(c2,c3|c1)$.

**Reply by josePereiro** [reply]
Estoy de acuerdo con lo que explicas, pero creo no es lo mismo que yo digo, aunque muy parecido... Yo solo quiero crear una intuición sobre la relación entre la información que está en los datos y la arquitectura de las redes. 

Yo parto de la idea de que tengo un texto que separo en todos sus triples y calculo sus frecuencias, obteniendo $P(c1, c2, c3)$. Hasta aquí, no hay machine learning. 

Ahora, yo voy a tener dos redes que las entreno siempre con los mismos triples (literalmente), solo que de maneras diferentes. Una la entreno para generar un próximo token dado dos previos , y otra para generar dos tokens dado uno anterior. Again, con los mismos triples, solo cambio quien es label y quien es input. Mi punto es que, si ambas redes son entrenadas con los mismos triples, en ambos casos solo estamos codificando, somehow, el tensor $P(c1, c2, c3)$ inicial... La diferencia entre $P(c1|c2, c3)$ y $P(c1, c2| c3)$ es solo una cuestión de arquitectura. Fijate que si todo funciona en el entrenamiento, el texto que es generado debe tener una estadísticas de sus triples parecida al los triples originales. 

Otra manera de verlo es que una vez generado el texto no puedes saber de qué red fue generado, que fuera dado uno dame dos, o dado dos dame uno no influye. Es solo un texto sampleado desde $P(c1, c2, c3)$.

**Reply by josePereiro** [reply]
Seguro se puede entrenar una red que genere los triples sin input, lo que ahora no me imagino como puede ser la _loss function_.

**Reply by Fourier92** [reply]
Man, es la misma conclusión a la q me refería anteriormente. No importa si tú matriz inicial, P(c1, c2, c3), es la misma para todos. Lo q es importante es la pregunta que le haces a esta matriz. Y en cada caso estás haciendo una pregunta diferente que da como resultado distribuciones de probabilidad diferentes. La frecuencia de aparición de los caracteres diferirá en ambos casos y ambos difieren de la matriz inicial.

**Reply by josePereiro** [reply]
> La frecuencia de aparición de los caracteres diferirá en ambos casos y ambos difieren de la matriz inicial.

Digamos que tenemos una lista con 1000 triplets y $abc$ aparatece 100 veces. Podemos preguntar, sobre esos mismos triplets, cuantas veces se termina en $c$ dado que el triplet empieza con $ab$?, serán 100 veces. También podemos preguntarnos, cuantas veces aparece $bc$ en los triplets que empiezan con $a$?, serán 100 veces. En todos los casos nos referimos a los mismos triplets $abc$. Si entrenamos las redes en cuestión minimizando el error con la misma lista de triplets, estas deben generar texto que tenga ~ 100 $abc$ triplets por cada 1000. 

No importa el tipo de probabilidades internas que usemos, para minimizar el error estas deben ser consistentes con los datos. Formalmente, lo que debe pasar es que $P(c1,c2,c3) \propto P(c1| c2, c3) \propto P(c1, c2| c3) \propto N(c1, c2, c3)$ donde $N(c1,c2,c3)$ es la cantidad de veces que ese triplet  aparece en los datos. Fíjate que no es que las probabilidades son iguales, son proporcionales... Pero eso es suficiente, por ejemplo, en el caso del modelo que genere basado en la probabilidad más alta, $max(P(c1,c2,c3)) == max(P(c1| c2, c3)) max(P(c1, c2| c3))$, por lo que literalmente se generaran los mismos strings.
